### PR TITLE
[extension/healthcheckv2extension] Refactor internal/status to pkg/status

### LINKED ISSUE #34692: [extension/healthcheckv2] Ability to access the `status.Aggregator` from another extension
### Component(s)

extension/healthcheckv2

### Is your feature request related to a problem? Please describe.

Currently the `opampextension` uses the original healthcheckextension and it determines if the collector is healthy by performing an HTTP request to its own process to determine if the collector is health. https://github.com/open-telemetry/opamp-go/blob/ed38d5f4bf930b57e04581919fbfb676aaa5a5af/internal/examples/supervisor/supervisor/supervisor.go#L435

This is not great from the standpoint of requiring the healthcheck endpoint to be exposed on the host just to get health of the current process that the `opampextension` is also running in. If you don't want any endpoint exposed then `opampextension` is unable to determine health. It is also very inefficient to determine status as it requires the a HTTP connection from its own process to the same process to just get status information. Another possible issue is IP table rules that prevent localhost port access prevent the ability to get this information (rare, but I have see it happen). It would be much better if this could be done in process.

It would be better if an extension could access the `healthcheckv2extension` internal `status.Aggregator` so it can determine the status of the collector without having to perform HTTP/GRPC requests and expose an endpoint on the host. 

### Describe the solution you'd like

It would be great if something like the following could be done from inside another extension to access the aggregator:

```
func (hc *exampleExtension) Start(ctx context.Context, host component.Host) error {
	extensions := host.GetExtensions()
	component, ok := extensions[component.MustNewID("healthcheckv2")]
	if !ok {
		return fmt.Errorf("failed to find healthcheckv2 extension; required dependent")
	}
	aggregatorAccess, ok := component.(healthcheckv2extension.AggregatorAccess)
	if !ok {
		return fmt.Errorf("failed to find healthcheckv2 extension; unable to convert to AggregatorAccess interface")
	}
	aggregator := aggregatorAccess.GetAggregator()
	// can now call aggregator.AggregateStatus or aggregator.Subscribe
	return nil
}
```

Another option would just to expose `AggregateStatus` and `Subscribe` directly on the `healthcheckv2extension` through an interface without exposing the entire function interface to the caller because `RecordStatus` should not really be called externally.

### Describe alternatives you've considered

Another alternative would allow an extension to register a sub-component inside of the `healthcheckv2extension`. This would allow another extension to expose a different way of getting the health check information versus only HTTP and/or GRPC. This could be done using the same approach:

```
func (hc *exampleExtension) Start(ctx context.Context, host component.Host) error {
	extensions := host.GetExtensions()
	component, ok := extensions[component.MustNewID("healthcheckv2")]
	if !ok {
		return fmt.Errorf("failed to find healthcheckv2 extension; required dependent")
	}
	addComponent, ok := component.(healthcheckv2extension.AddSubcomponentInterface)
	if !ok {
		return fmt.Errorf("failed to find healthcheckv2 extension; unable to convert to AddSubcomponentInterface interface")
	}
        component := createNewSubComponent(ctx)
	return addComponent.AddComponent(component)
}
```

Downside of this approach is if the extension sets the `healthcheckv2extension` as a `Dependent()` then it will already be started (I believe) before `Start(...)` is called on the extension. The `AddComponent` function would then need to start the sub-component at the time of registering. I also believe this will require exposing more internal interfaces and types to extensions that might not be desired.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by blakerouse**
#### Description

Refactors the `extension/healthcheckv2extension/internal/status` into `pkg/status`.

This exposes the aggregator to be used by other extensions to gather component status information using the `extension.StatusWatcher`.

#### Link to tracking issue

Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34692

#### Testing

Being it was a refactor and all the same tests provided coverage, no additional testing was added.

#### Documentation

Added a `README.md` to the `pkg/status` to provide information on where this package can be used.

### CODE REVIEW

**Review by mwear — COMMENTED** [review]
Thanks for taking on this refactor. I made [a PR](https://github.com/blakerouse/opentelemetry-collector-contrib/pull/1) to this PR with a few additional updates that will be needed. In addition to those, this should have a changelog entry. Run `make chlog-new` to generate one and fill in the details.

File: pkg/status/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# pkg/status
+
+## Overview
+
+This module contains functionality for aggregating status information about components. Used by extensions that
+implement the `extension.StatusWatcher` interface for collecting status information about the internal status of all
```
Comment (mwear) [inline pkg/status/README.md:6]:
```suggestion
implement the `componentstatus.Watcher` interface for collecting status information about the internal status of all
```

The interface was renamed on core. I suspect there are some stale references in docs / comments that need to be updated.

**Comment by blakerouse** [comment]
@mwear Thanks for the cleanup PR. I have merged your PR and updated the `README.md` to the updated interface name.

**Comment by blakerouse** [comment]
@mwear Forgot about the changelog. I have added that as well.

File: pkg/status/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+## Overview
+
+This module contains functionality for aggregating status information about components. Used by extensions that
+implement the `componentstatus.Watcher` interface for collecting status information about the internal status of all
+components inside the collector.
+
+### History
+
+`status` was originally developed for the `healthcheckv2extension`. Other extensions that need to collector status
```
Comment (mwear) [inline pkg/status/README.md:11]: Can you double check the wording on this second sentence?

**Review by mwear — APPROVED** [review]
LGTM, thanks @blakerouse!

cc: @bacherfl, @evan-bradley re: #35488.

File: .chloggen/pkg-status.yaml
Comment (evan-bradley) [inline .chloggen/pkg-status.yaml:1]: I think this should be a `new_component` changelog announcing the `pkg/status` package. The refactoring of the healthcheckv2extension is essentially an internal change.

File: pkg/status/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# pkg/status
+
```
Comment (evan-bradley) [inline pkg/status/README.md:2]:
```suggestion
<!-- status autogenerated section -->
<!-- end autogenerated section -->
```
We should include an autogenerated section in here that will be filled out using the information from `metadata.yaml` when `make -C pkg/status generate` is run.

File: pkg/status/metadata.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+status:
```
Comment (evan-bradley) [inline pkg/status/metadata.yml:1]:
```suggestion
status:
  class: pkg
  stability:
    development: [ extension ]
```
We should include a few more details here just for completeness.

**Review by evan-bradley — COMMENTED** [review]
Thank you @blakerouse!

Overall this looks good to me, I just have comments on a few supplementary details.

**Comment by blakerouse** [comment]
@evan-bradley I have update the requested items.

But running `mage -C pkg/status generate` resulted in:

```
make: *** No rule to make target `generate'.  Stop.
```

**Comment by mwear** [comment]
> But running `make -C pkg/status generate` resulted in:
> 
> ```
> make: *** No rule to make target `generate'.  Stop.
> ```

I'm not sure why that doesn't work for a subdirectory, but you can run `make generate` from the root and will run for each contrib component. It's slow, but it will get the job done.

**Comment by blakerouse** [comment]
@mwear I did do that. I didn't see it generate or do anything. Maybe there is nothing to generate?

**Comment by mwear** [comment]
> @mwear I did do that. I didn't see it generate or do anything. Maybe there is nothing to generate?

It looks like there are two more changes needed:

Add a doc.go file with the following contents:

```
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

//go:generate mdatagen metadata.yaml

package status // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/status"

```

Update metadata.yml to include a type, you can just copy this:

```
type: status

status:
  class: pkg
  stability:
    development: [extension]
  codeowners:
    active: [jpkrohling, mwear]
```

I found that if you cd into the pkg/status directory you can run `mdatagen metadata.yml` to generate for a specific component. `make generate` from the root should also work after these changes.

**Comment by blakerouse** [comment]
@mwear @evan-bradley Thanks for the help on getting this ready. I think its ready now as the generation logic worked this time and filled in the correct information in the README.

**Comment by mwear** [comment]
The currently failing CI check says to run `make generate-gh-issue-templates` and commit the changes.

**Comment by blakerouse** [comment]
@mwear Okay. maybe now its ready. 😄

File: pkg/status/README.md (hunk 2 above)
Comment (evan-bradley) [reply pkg/status/README.md:21]: Not a blocker, but could you take a look at this @blakerouse?

**Review by mwear — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Comment by blakerouse** [comment]
No problem! Glad to help.

**Pull request merged**